File: backend/app/schemas/student.py
"""Student related schemas."""
from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


class StudentBase(BaseModel):
    """Base student schema."""
    email: str
    name: str | None = None
    phone: str | None = None


class StudentCreate(StudentBase):
    """Student creation schema."""
    pass


class StudentUpdate(BaseModel):
    """Student update schema."""
    name: str | None = None
    phone: str | None = None


class StudentResponse(StudentBase):
    """Student response schema."""
    id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeBase(BaseModel):
    """Base resume schema."""
    filename: str | None = None
    file_type: str | None = None
    is_primary: bool = False


class ResumeCreate(ResumeBase):
    """Resume creation schema."""
    student_id: UUID
    file_path: str | None = None


class ResumeUpdate(BaseModel):
    """Resume update schema."""
    filename: str | None = None
    is_primary: bool | None = None


class ResumeResponse(ResumeBase):
    """Resume response schema."""
    id: UUID
    student_id: UUID
    file_path: str | None = None
    raw_text: str | None = None
    parsed_data: dict[str, Any] | None = Field(None, validation_alias="parsed_json")
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True, populate_by_name=True)


class StudentProfileBase(BaseModel):
    """Base student profile schema."""
    skills: list[str] | None = None
    experience: list[dict[str, Any]] | None = None
    education: list[dict[str, Any]] | None = None
    projects: list[dict[str, Any]] | None = None
    interests: list[str] | None = None
    goals: list[str] | None = None


class StudentProfileCreate(StudentProfileBase):
    """Student profile creation schema."""
    student_id: UUID


class StudentProfileUpdate(BaseModel):
    """Student profile update schema."""
    profile_json: dict[str, Any] | None = None


class StudentProfileResponse(BaseModel):
    """Student profile response schema."""
    id: UUID
    student_id: UUID
    profile_json: dict[str, Any]
    completeness_score: float = 0.0
    evidence_json: dict[str, Any] | None = None
    version: str = "1.0"
    missing_suggestions: list[str] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class ResumeUploadResponse(BaseModel):
    """Response for resume upload with parsing results."""
    resume: ResumeResponse
    parsed_data: dict[str, Any]
    completeness_score: float
    missing_suggestions: list[str]
    normalization_log: list[dict[str, str]]

File: frontend/src/types/student.ts
// Student types - aligned with backend/app/schemas/student.py

// Profile JSON structure
export interface ProfileBasicInfo {
  name?: string;
  email?: string;
  phone?: string;
  school?: string;
  degree?: string;
  major?: string;
}

export interface ProfileSkill {
  name: string;
  category: string;
  level?: string;
}

export interface ProfileExperience {
  type: 'project' | 'internship';
  title: string;
  company?: string;
  duration?: string;
  description?: string;
}

export interface ProfileEducation {
  school: string;
  degree: string;
  major?: string;
  duration?: string;
}

export interface ProfileCertificate {
  name: string;
  issuer: string;
  date?: string;
}

export interface ProfileAward {
  name: string;
  level?: string;
  date?: string;
  evidence?: string;
}

export interface ProfileJson {
  competitiveness_score?: number;
  experience_months?: number;
  basic_info?: ProfileBasicInfo;
  skills?: ProfileSkill[];
  experiences?: ProfileExperience[];
  certificate_names?: string[];
  education?: ProfileEducation[];
  certificates?: ProfileCertificate[];
  awards?: ProfileAward[];
  soft_skills?: Array<{
    dimension: string;
    score: number;
    evidence?: string;
  }> | Record<string, number>;
}

export interface StudentBase {
  email: string;
  name?: string;
  phone?: string;
}

export interface StudentCreate extends StudentBase {
  // nothing extra
}

export interface StudentUpdate {
  name?: string;
  phone?: string;
}

export interface StudentResponse extends StudentBase {
  id: string;
  created_at: string;
  updated_at: string;
}

export interface ResumeBase {
  filename?: string;
  file_type?: string;
  is_primary: boolean;
}

export interface ResumeCreate extends ResumeBase {
  student_id: string;
  file_path?: string;
}

export interface ResumeUpdate {
  filename?: string;
  is_primary?: boolean;
}

export interface ResumeResponse extends ResumeBase {
  id: string;
  student_id: string;
  file_path?: string;
  raw_text?: string;
  parsed_data?: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

export interface StudentProfileUpdate {
  profile_json?: Record<string, unknown>;
}

export interface StudentProfileResponse {
  id: string;
  student_id: string;
  profile_json: ProfileJson;
  completeness_score: number;
  evidence_json?: Record<string, unknown>;
  version: string;
  missing_suggestions?: string[];
  created_at: string;
  updated_at: string;
}

export interface ResumeUploadResponse {
  resume: ResumeResponse;
  parsed_data: Record<string, unknown>;
  completeness_score: number;
  missing_suggestions: string[];
  normalization_log: Array<{ original: string; normalized: string }>;
  parse_meta?: {
    status?: string;
    is_fallback?: boolean;
    retrying?: boolean;
  };
}

// Student Profile types
export interface StudentProfileBase {
  skills?: string[];
  experience?: Record<string, unknown>[];
  education?: Record<string, unknown>[];
  projects?: Record<string, unknown>[];
  interests?: string[];
  goals?: string[];
}

export interface StudentProfileCreate extends StudentProfileBase {
  student_id: string;
}
